testdata/p4_16_samples/pna-mux-dismantle.p4

```diff
@@ -0,0 +1,216 @@
+    bit<16> hdrChecksum;
+    bit<32> srcAddr;
+    bit<32> dstAddr;
+}
+
+header tcp_t {
+    bit<16> srcPort;
+    bit<16> dstPort;
+    bit<32> seqNo;
+    bit<32> ackNo;
+    bit<4>  dataOffset;
+    bit<3>  res;
+    bit<3>  ecn;
+    bit<6>  flags;
+    bit<16> window;
+    bit<16> checksum;
+    bit<16> urgentPtr;
+}
+
+
+struct empty_metadata_t {
+}
+
+typedef bit<48> ByteCounter_t;
+typedef bit<32> PacketCounter_t;
+typedef bit<80> PacketByteCounter_t;
+
+const bit<32> NUM_PORTS = 4;
+
+
+//////////////////////////////////////////////////////////////////////
+// Struct types for holding user-defined collections of headers and
+// metadata in the P4 developer's program.
+//
+// Note: The names of these struct types are completely up to the P4
+// developer, as are their member fields, with the only restriction
+// being that the structs intended to contain headers should only
+// contain members whose types are header, header stack, or
+// header_union.
+//////////////////////////////////////////////////////////////////////
+
+struct main_metadata_t {
+    bit<1> rng_result1;
+}
+
+// User-defined struct containing all of those headers parsed in the
+// main parser.
+struct headers_t {
+    ethernet_t ethernet;
+    ipv4_t ipv4;
+    tcp_t tcp;
+}
+
+control PreControlImpl(
+    in    headers_t  hdr,
+    inout main_metadata_t meta,
+    in    pna_pre_input_metadata_t  istd,
+    inout pna_pre_output_metadata_t ostd)
+{
+    apply {
+    }
+}
+
+parser MainParserImpl(
+    packet_in pkt,
+    out   headers_t       hdr,
+    inout main_metadata_t main_meta,
+    in    pna_main_parser_input_metadata_t istd)
+{
+    state start {
+        pkt.extract(hdr.ethernet);
+        transition select(hdr.ethernet.etherType) {
+            0x0800: parse_ipv4;
+            default: accept;
+        }
+    }
+    state parse_ipv4 {
+        pkt.extract(hdr.ipv4);
+        transition accept;
+    }
+}
+
+bit<1> dummy_func(inout headers_t hdr,   bit<16> min1,
+        bit<16> max1,inout main_metadata_t user_meta){
+    return  (bit<1>) ((min1 <= hdr.tcp.srcPort) && (hdr.tcp.srcPort <= max1));
+}
+
+action do_range_checks_0 (inout headers_t hdr,inout main_metadata_t user_meta,
+        bit<16> min1,
+        bit<16> max1) {
+    user_meta.rng_result1  = dummy_func(hdr,min1, max1, user_meta );
+}
+
+control MainControlImpl(
+    inout headers_t       hdr,           // from main parser
+    inout main_metadata_t user_meta,     // from main parser, to "next block"
+    in    pna_main_input_metadata_t  istd,
+    inout pna_main_output_metadata_t ostd)
+{
+    action next_hop(PortId_t vport) {
+        send_to_port(vport);
+    }
+    action add_on_miss_action() {
+        bit<32> tmp = 0;
+        add_entry(action_name="next_hop", action_params = tmp);
+    }
+
+    action do_range_checks_1 (
+        bit<16> min1,
+        bit<16> max1)
+    {
+        user_meta.rng_result1 =  (bit<1>) ((min1 <= hdr.tcp.srcPort) && (hdr.tcp.srcPort <= max1));
```

Comment:
Is the nested expression supported? If yes, that can be added as additional test.

Reply:
As in above comment writing ternary directly in p4, looks like translated by frontend itself, Ternary introduced by LowerExpressions again needs to be translated to if else chain.

But I tried adding something like this 
        user_meta.rng_result1 = ((min1 <= hdr.tcp.srcPort) && (hdr.tcp.srcPort <= max1)) ? (bit<1>)((16w50 <= hdr.tcp.srcPort) && (hdr.tcp.srcPort <= 16w100)) : user_meta.val1;

And it compiles ok with this patch.

Reply:
By nested I mean like this: a > b ? c : ((a > d) ? d : a)

Reply:
Updated testcase with nested ternary. 
Although it passed without any changes, copied missing pieces from DoSimplifyExpression to make it even.